File: Libraries/Http/native/java-app/source/HttpHelper.java
package crayonlib.http;

import org.crayonlang.interpreter.structs.*;

import java.io.DataOutputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class HttpHelper {

	public static void readResponseData(
		Object httpRequest,
		int[] intOut,
		String[] stringOut,
		Object[] responseNativeData,
		ArrayList<String> headersOut) {
		
		HttpResponseValue response = (HttpResponseValue)httpRequest;
		intOut[0] = response.statusCode;
		intOut[1] = response.isContentBinary ? 1 : 0;
		stringOut[0] = response.statusMessage;
		stringOut[1] = response.contentString;
		responseNativeData[0] = response.contentBytes;
		for (int i = 0; i < response.headers.length; ++i) {
			headersOut.add(response.headers[i]);
		}
	}

	public static void getResponseBytes(
		Object byteArrayObj,
		Value[] integersCache,
		ArrayList<Value> output) {
		
		byte[] bytes = (byte[])byteArrayObj;
		int length = bytes.length;
		for (int i = 0; i < length; ++i)
		{
			output.add(integersCache[bytes[i]]);
		}
	}

	public static void sendRequestAsync(
		final Object[] requestNativeData,
		String method,
		String url,
		ArrayList<String> headers,
		int contentMode, // 0 - null, 1 - text, 2 - binary
		Object content,
		boolean outputIsBinary,
		VmContext vmContext,
		Value callbackFp,
		Object[] mutexStuff) {
		
		requestNativeData[1] = new Object();
		
		new Thread(new Runnable() {
			public void run() {
				HttpResponseValue response = sendRequestSyncImpl(requestNativeData, method, url, headers, contentMode, content, outputIsBinary);
				synchronized (requestNativeData[1]) {
					requestNativeData[0] = response;
					requestNativeData[2] = true;
				}
				
				synchronized(mutexStuff[0]) {
					((ArrayList<Value>)mutexStuff[1]).add(callbackFp);
				}
			}
		}).start();
	}

	public static boolean sendRequestSync(
		Object[] requestNativeData,
		String method,
		String url,
		ArrayList<String> headers,
		int contentMode, // 0 - null, 1 - text, 2 - binary
		Object content,
		boolean outputIsBinary) {
		
		HttpResponseValue response = sendRequestSyncImpl(requestNativeData, method, url, headers, contentMode, content, outputIsBinary);
		requestNativeData[0] = response;
		requestNativeData[2] = true;
		return false;
	}

	private static HttpResponseValue sendRequestSyncImpl(
		Object[] requestNativeData,
		String method,
		String urlString,
		ArrayList<String> headers,
		int contentMode, // 0 - null, 1 - text, 2 - binary
		Object content,
		boolean outputIsBinary) {
		
		String responseString = null;
		byte[] responseBytes = null;
		int statusCode = 0;
		String statusMessage = "";
		ArrayList<String> responseHeaders = new ArrayList<>();
		
		HttpURLConnection connection = null;
		try {
			URL url = new URL(urlString);
			connection = (HttpURLConnection) url.openConnection();
			connection.setRequestMethod(method);
			boolean contentLengthSet = false;
			for (int i = 0; i < headers.size(); i += 2) {
				String headerName = headers.get(i);
				String headerValue = headers.get(i + 1);
				connection.setRequestProperty(headerName, headerValue);
				if (headerName.toLowerCase().equals("content-length")) {
					contentLengthSet = true;
				}
			}
			// TODO: verify this isn't automatically done for you.
			if (!contentLengthSet && contentMode != 0) {
				int length;
				if (contentMode == 1) {
					length = ((String)content).length();
				} else {
					length = ((byte[])content).length;
				}
				connection.setRequestProperty("Content-Length", "" + length);
			}
			
			if (contentMode != 0) {
				DataOutputStream wr = new DataOutputStream(connection.getOutputStream());
				if (contentMode == 1) {
					wr.writeBytes((String) content);
				} else {
					throw new RuntimeException("Bytes not implemented yet.");
				}
				wr.close();
			}
			
			InputStream is = connection.getInputStream();
			
			if (outputIsBinary) {
				java.io.ByteArrayOutputStream outputStream = new java.io.ByteArrayOutputStream();
				int bytesRead = 1;
				byte[] buffer = new byte[1024];
				while (bytesRead > -1) {
					bytesRead = is.read(buffer, 0, buffer.length);
					if (bytesRead > -1) {
						outputStream.write(buffer, 0, bytesRead);
					}
				}
				
				outputStream.flush();
				responseBytes = outputStream.toByteArray();
			} else {
				StringBuilder sb = new StringBuilder();
				char[] buffer = new char[1024];
				try {
					// TODO: something better than just assuming UTF-8
					Reader in = new InputStreamReader(is, "UTF-8");
					while (true) {
						int charsRead = in.read(buffer, 0, buffer.length);
						if (charsRead < 0) {
							break;
						}
						sb.append(buffer, 0, charsRead);
					}
				} catch (Exception e2) {
					
				}
				responseString = sb.toString();
			}
			
			statusCode = connection.getResponseCode();
			statusMessage = connection.getResponseMessage();
			if (statusMessage == null) statusMessage = "";
			for (Map.Entry<String, List<String>> headerPair : connection.getHeaderFields().entrySet()) {
				String headerName = headerPair.getKey();
				if (headerName != null) {
					for (String headerValue : headerPair.getValue()) {
						responseHeaders.add(headerName);
						if (headerValue == null) headerValue = "";
						responseHeaders.add(headerValue);
					}
				}
			}
		} catch (Exception e1) {
			System.out.println(e1);
		} finally {
			if (connection != null) {
				connection.disconnect();
			}
		}
		
		HttpResponseValue output = new HttpResponseValue();
		output.statusCode = statusCode;
		output.statusMessage = statusMessage;
		output.isContentBinary = outputIsBinary;
		output.contentString = responseString;
		output.contentBytes = responseBytes;
		output.headers = responseHeaders.toArray(new String[0]);
		return output;
	}

	private static class HttpResponseValue
	{
		public int statusCode;
		public String statusMessage;
		public boolean isContentBinary;
		public String contentString;
		public byte[] contentBytes;
		public String[] headers;
	}

	public static boolean pollRequest(final Object[] requestNativeData) {
		boolean output = false;
		synchronized (requestNativeData[1]) {
			output = (boolean) requestNativeData[2];
		}
		return output;
	}
}

File: Libraries/Http/native/csharp-app/source/HttpHelper.cs
using System;
using System.Collections.Generic;
using System.Net;
using Interpreter.Structs;

namespace Interpreter.Libraries.Http
{
    internal static class HttpHelper
    {
        public static void ReadResponseData(
            object httpRequest,
            int[] intOut,
            string[] stringOut,
            object[] responseNativeData,
            List<string> headersOut)
        {
            HttpResponseValue response = (HttpResponseValue)httpRequest;
            intOut[0] = response.StatusCode;
            intOut[1] = response.IsContentBinary ? 1 : 0;
            stringOut[0] = response.StatusMessage;
            stringOut[1] = response.ContentString;
            responseNativeData[0] = response.ContentBytes;
            headersOut.AddRange(response.Headers);
        }

        public static void GetResponseBytes(object byteArrayObj, Value[] integersCache, List<Value> output)
        {
            byte[] bytes = (byte[])byteArrayObj;
            int length = bytes.Length;
            for (int i = 0; i < length; ++i)
            {
                output.Add(integersCache[bytes[i]]);
            }
        }

        public static void SendRequestAsync(
            object[] requestNativeData,
            string method,
            string url,
            List<string> headers,
            int contentMode, // 0 - null, 1 - text, 2 - binary
            object content,
            bool outputIsBinary,
            VmContext vmContext,
            Value callback,
            object[] mutexDataIgnored) // 
        {
            requestNativeData[1] = new object();
            System.ComponentModel.BackgroundWorker bgWorker = new System.ComponentModel.BackgroundWorker();
            bgWorker.DoWork += (sender, e) =>
            {
                HttpResponseValue response = SendRequestSyncImpl(requestNativeData, method, url, headers, contentMode, content, outputIsBinary);
                lock (requestNativeData[1])
                {
                    requestNativeData[0] = response;
                    requestNativeData[2] = true;
                }
            };
            bgWorker.RunWorkerCompleted += (sender, e) =>
            {
                Interpreter.Vm.CrayonWrapper.runInterpreterWithFunctionPointer(vmContext, callback, new Value[0]);
            };
            bgWorker.RunWorkerAsync();
        }

        public static bool SendRequestSync(
            object[] requestNativeData,
            string method,
            string url,
            List<string> headers,
            int contentMode, // 0 - null, 1 - text, 2 - binary
            object content,
            bool outputIsBinary)
        {
            HttpResponseValue response = SendRequestSyncImpl(requestNativeData, method, url, headers, contentMode, content, outputIsBinary);

            requestNativeData[0] = response;
            requestNativeData[2] = true;

            return false;
        }

        private static HttpResponseValue SendRequestSyncImpl(
            object[] requestNativeData,
            string method,
            string url,
            List<string> headers,
            int contentMode, // 0 - null, 1 - text, 2 - binary
            object content,
            bool outputIsBinary)
        {
            HttpWebRequest request = (HttpWebRequest)HttpWebRequest.Create(url);
            request.Method = method;
            string contentType = null;
            for (int i = 0; i < headers.Count; i += 2)
            {
                string headerName = headers[i];
                string headerValue = headers[i + 1];
                switch (headerName.ToLower())
                {
                    case "user-agent":
                        request.UserAgent = headerValue;
                        break;
                    case "content-type":
                        contentType = headerValue;
                        break;
                    default:
                        request.Headers.Add(headerName, headerValue);
                        break;
                }
            }

            if (contentMode != 0)
            {
                byte[] contentBytes;
                if (contentMode == 2)
                {
                    int[] intContent = (int[])content;
                    int length = intContent.Length;
                    contentBytes = new byte[length];
                    for (int i = 0; i < length; ++i)
                    {
                        contentBytes[i] = (byte)intContent[i];
                    }
                }
                else
                {
                    string body = (string)content;
                    contentBytes = System.Text.UTF8Encoding.UTF8.GetBytes(body);
                }
                request.ContentLength = Convert.ToInt64(contentBytes.Length);
                if (contentType != null)
                {
                    request.ContentType = contentType;
                }
                System.IO.BinaryWriter streamWriter = new System.IO.BinaryWriter(request.GetRequestStream());
                streamWriter.Write(contentBytes);
                streamWriter.Flush();
                streamWriter.Close();
            }

            HttpWebResponse response;
            try
            {
                response = (HttpWebResponse)request.GetResponse();
            }
            catch (WebException we)
            {
                response = (HttpWebResponse)we.Response;
                if (response == null && we.Status == WebExceptionStatus.Timeout)
                {
                    // TODO: timeouts
                    throw new Exception();
                }
            }

            int statusCode = (int)response.StatusCode;
            string statusMessage = response.StatusDescription;
            byte[] responseBytes = null;
            string responseString = null;
            if (outputIsBinary)
            {
                System.IO.BinaryReader binaryReader = new System.IO.BinaryReader(response.GetResponseStream());
                byte[] buffer = new byte[1024];
                List<byte> contentList = new List<byte>();
                int bytesRead = 1;
                while (bytesRead > 0)
                {
                    bytesRead = binaryReader.Read(buffer, 0, buffer.Length);
                    if (bytesRead == 1024)
                    {
                        contentList.AddRange(buffer);
                    }
                    else
                    {
                        for (int i = 0; i < bytesRead; ++i)
                        {
                            contentList.Add(buffer[i]);
                        }
                    }
                }
                binaryReader.Close();
                responseBytes = contentList.ToArray();
            }
            else
            {
                System.IO.StreamReader streamReader = new System.IO.StreamReader(response.GetResponseStream());
                responseString = streamReader.ReadToEnd();
                streamReader.Close();
            }

            List<string> responseHeaders = new List<string>();
            foreach (string headerName in response.Headers.AllKeys)
            {
                foreach (string headerValue in response.Headers.GetValues(headerName))
                {
                    responseHeaders.Add(headerName);
                    responseHeaders.Add(headerValue);
                }
            }

            if (outputIsBinary && statusCode >= 400)
            {
                try
                {
                    responseString = System.Text.Encoding.UTF8.GetString(responseBytes);
                }
                catch (Exception)
                { }
            }

            return new HttpResponseValue()
            {
                StatusCode = statusCode,
                StatusMessage = statusMessage,
                IsContentBinary = outputIsBinary,
                ContentString = responseString,
                ContentBytes = responseBytes,
                Headers = responseHeaders.ToArray(),
            };
        }

        private class HttpResponseValue
        {
            public int StatusCode { get; set; }
            public string StatusMessage { get; set; }
            public bool IsContentBinary { get; set; }
            public string ContentString { get; set; }
            public byte[] ContentBytes { get; set; }
            public string[] Headers { get; set; }
        }

        public static bool PollRequest(object[] requestNativeData)
        {
            bool output = false;

            lock (requestNativeData[1])
            {
                output = (bool)requestNativeData[2];
            }

            return output;
        }
    }
}
